KIPs/KIP-277.md

```diff
@@ -0,0 +1,254 @@
+
+    /// @notice Cancel existing request for the given consensus node id
+    /// @param consensusNodeId The consensus node id that the request is for
+    /// @dev Callable by validator manager
+    function cancelRequest(address consensusNodeId) external;
+
+    /// @notice Transfer the validator manager to a new address
+    /// @dev Callable by validator manager
+    ///      Can't transfer manager if there is a pending request
+    function transferManager(address consensusNodeId, address newManager) external;
+
+    /// @notice Approve a request
+    /// @dev Callable by owner
+    /// @param consensusNodeId The consensus node id that has a request to be approved
+    function approveRequest(address consensusNodeId, uint256 requestIndex) external;
+
+    /// @notice Reject request of a validator manager
+    /// @dev Callable by owner
+    /// @param consensusNodeId The consensus node id that has a request to be rejected
+    function rejectRequest(address consensusNodeId, uint256 requestIndex) external;
+}
+```
+
+After the activation of `VMC`, all validators must manage their own registration process through `VMC`. For onboarding of existing validators, the Kaia team will initially backfill the validator information at the deployment of `VMC`.
+
+### Prerequisites
+
+Since the `VMC` will execute calls to the `AddressBook` and `SimpleBlsRegistry`, the following prerequisites should be satisfied to ensure the execution of requests:
+
+1. `VMC` is registered as an admin of `AddressBook`. It can include more than one admin for backup purposes.
+2. `AddressBook` has `requirement` set to 1.
+   - This is required since `AddressBook` internally implements a multisig mechanism for the admin list.
+3. `VMC` is the owner of `SimpleBlsRegistry`.
+
+If any of the prerequisites are not met, the `VMC` MUST revert the request approval. (not prevent request submission)
+
+### Self Registration Process
+
+#### Onboarding
+
+1. Candidate manager deploys a new consensus staking contract (`CnStakingV3MultiSig`) through the registered staking factory.
+2. Candidate finishes the initialization of staking contract.
+3. Candidate submits an onboarding request to `VMC`.
+4. Kaia team reviews the request.
+5. Kaia team approves the request.
+   - a. `VMC` MUST execute `AddressBook.submitRegisterCnStakingContract` to register the validator information.
+   - b. `VMC` MUST execute `SimpleBlsRegistry.register` to register the BLS public key information.
+
+#### Offboarding
+
+1. Validator manager submits an offboarding request to `VMC`.
+2. Kaia team reviews the request.
+3. Kaia team approves the request.
+   - a. `VMC` MUST execute `AddressBook.submitUnregisterCnStakingContract` to unregister the validator information of staking contracts.
+   - b. `VMC` MUST execute `SimpleBlsRegistry.unregister` to unregister the BLS public key information.
+
+#### AddStakingContract
```

Comment:
`AddStakingContract` spam control suggestion
I think unlimited `AddStakingContract` requests may create potential problems.
What do you think about adding some methods to reduce problems?

1) Per-validator cap on staking contracts (for example, up to 3 contracts)
2) Minimum stake per new contract (for example, at least 500k KAIA), with rewards eligible only after the threshold is met

Reply:
Thank you for the suggestion.

That's a valid concern once we enter into permissionless. We MUST implement proper staking-related validations to prevent any spamming or misbehavior. But for now, KIP-277 focuses more on the current operation compatibility while making stepwise changes, so let those validations be offchain would be flexible.

Add rationale [dd4b78b](https://github.com/kaiachain/kips/pull/78/commits/dd4b78b6c4bc98edb5203c6025e832621e2a26de).

Reply:
I see. I understood that KIP-277 is temporary and will be replaced when we move toward permissionless.
Thank you for the clarification.